src/nnet3/nnet-simple-component.cc

```diff
@@ -295,6 +295,110 @@ void NoOpComponent::Backprop(const std::string &debug_info,
+  std::stringstream stream;
+  stream << Type() << ", dim=" << dim_
+         << ", norm-based-clipping=" << (norm_based_clipping_ ? "true" : "false")
+         << ", clipping-threshold=" << clipping_threshold_;
+  return stream.str();
+}
+
+void GradientClipperComponent::Init(int32 dim, 
+                                    BaseFloat clipping_threshold,
+                                    bool norm_based_clipping)  {
+  KALDI_ASSERT(clipping_threshold >= 0 && dim > 0);
+  dim_ = dim;
+  norm_based_clipping_ = norm_based_clipping;
+  clipping_threshold_ = clipping_threshold;
+}
+
+void GradientClipperComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  bool ok = cfl->GetValue("dim", &dim);
+  bool norm_based_clipping = false;
+  BaseFloat clipping_threshold = 15.0;
+  cfl->GetValue("clipping-threshold", &clipping_threshold);
+  cfl->GetValue("norm-based-clipping", &norm_based_clipping);
+  if (!ok || cfl->HasUnusedValues() ||
+      clipping_threshold < 0 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, clipping_threshold, norm_based_clipping);
+}
+
+void GradientClipperComponent::Propagate(
+                                 const ComponentPrecomputedIndexes *indexes,
+                                 const CuMatrixBase<BaseFloat> &in,
+                                 CuMatrixBase<BaseFloat> *out) const {
+  out->CopyFromMat(in);
+}
+
+void GradientClipperComponent::Backprop(const std::string &debug_info,
+                             const ComponentPrecomputedIndexes *indexes,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &,
+                             const CuMatrixBase<BaseFloat> &out_deriv,
+                             Component *to_update, // may be NULL; may be identical
+                             // to "this" or different.
+                             CuMatrixBase<BaseFloat> *in_deriv) const {
+  // the following statement will do nothing if in_deriv and out_deriv have same
+  // memory.
+  in_deriv->CopyFromMat(out_deriv);
+  BaseFloat norm_before_clipping = 0,
+            norm_after_clipping = 0;
+
+  if (GetVerboseLevel() >= 2) {
+    norm_before_clipping = out_deriv.FrobeniusNorm();
+  }
+  if (clipping_threshold_ > 0) {
+    if (norm_based_clipping_) {
+      // each row in the derivative matrix, which corresponds to one sample in
+      // the mini-batch, is scaled to have a max-norm of clipping_threshold_
+      BaseFloat clipping_threshold_2 = pow(clipping_threshold_, 2);
+      CuVector<BaseFloat> deriv_norm(in_deriv->NumRows());
+      deriv_norm.AddDiagMat2(1.0 / in_deriv->NumCols(),
+                             *in_deriv, kNoTrans, 0.0);
+      deriv_norm.ApplyFloor(clipping_threshold_2);
+      deriv_norm.Scale(1.0 / clipping_threshold_2);
+      deriv_norm.ApplyPow(-0.5);
+      in_deriv->MulRowsVec(deriv_norm);
```

Comment:
If you make this a bit more like the code I sketched by email, here, it will be a bit more efficient:
   CuVector<BaseFloat> clipping_scales(in_deriv->NumRows());
    clipping_scales.AddDiagMat2(pow(clipping_threshold_, -2), *in_deriv, kNoTrans, 0.0);
    // now clipping_scales contains the squared (norm of each row divided by clipping_threshold)
    int32 num_not_scaled = clipping_scales.ApplyFloor(1.0); 
    // now clipping_scales contains min(1, squared-(norm/clipping_threshold))
    if (num_not_scaled != clipping_scales.Dim()) {
      clipping_scales.ApplyPow(-0.5); 
      // now clipping_scales contains max(1, clipping_threshold/vector_norm)
      in_deriv->MulRowsVec(clipping_scales);
    }
I notice you use a scale of 1.0 / in_deriv->NumCols().  This way the clipping threshold corresponds not to a vector norm but to a vector norm per element.  I think using a vector norm might make more sense (i.e. skipping the 1.0/in_deriv->NumCols() factor), because for higher-dimensional features, the derivative per element might be smaller because each element matters less.  Also, using an overall vector norm is just simpler to explain.
Also, you can use the num_not_scaled value to keep track of what proportion of vectors were scaled, for stats.  You just need to store two scalars.  Please implement ZeroStats() to set them to zero.  I'll be changing the documentation for ZeroStats() to indicate that it's not just for components that implement StoreStats().